pkg/urbit/vere/pier.c

```diff
@@ -707,11 +524,33 @@ _pier_work_save(u3_pier* pir_u)
   u3_disk* log_u = pir_u->log_u;
   u3_save* sav_u = pir_u->sav_u;
 
-  c3_assert( god_u->dun_d == sav_u->req_d );
+  // This is wrong?  We can have requested that the worker should snapshot
+  // between when the worker was given a compute request and when the worker
+  // returns a compute response. (See the c3_max() call in u3_pier_snap().) I
+  // don't know what the semantics are supposed to be here.
+  //
+  // Let's say we're in replay of the event log from nothing. We might have
+  // u3_pier_snap() called from a timer. I've seen crashes when:
+  //
+  //   god_u->dun_d: 24, sav_u->req_d: 25, log_u->com_d: 31
+  //
+  // com_d is greater than req_d is greater than dun_d because we're waiting
+  // for the worker to complete replay of already committed events. And this
+  // seems OK because we can totally request that the worker produce a snapshot
+  // of event 25 even though we haven't marked it as done because we know that
+  // it committed previously?
+  //
+  // No! This breaks because it ignores the req_d and instead tells the worker
+  // to go with what was req_d.
+  //
+  // c3_assert( god_u->dun_d == sav_u->req_d );
+
```

Comment:
I'd like to draw the reviewer's attention to this big comment block to double check my reasoning. If that plus the TODO right under this are correct, I'll pull this comment block and the wrong assertion out.

Reply:
Right now, the worker is only able to take a snapshot of the state derived from the latest event. So this assertion as written is actually correct (and there's an equivalent assertion in the worker code as well). If you've seen failures in this assertion, it suggests that this function is being called out of order, or some of the bookkeeping in pier.c is not being performed directly. I'm not sure exactly what might be going wrong, but I'm not comfortable with removing this assertion.

Reply:
At any point, the u3_save_io_init() timer can fire, including between when we've requested the worker to perform some computation and when we hear back asynchronously from the worker. The way I solved this before disabling the assertion above was to comment out u3_save_io_init(). I'm pretty sure this assertion is broken, even without the lmdb changes.

Reply:
This is a good point, I hadn't anticipated the timer firing during replay while in the middle of computing an event. I think the right fix is to change the condition gating the `_pier_work_save()` call inside `u3_pier_snap()` to something like:

```
if ( (log_u->com_d >= top_d) &&
          (god_u->dun_d == top_d) )
{
...
```

In other words, only immediately call `_pier_work_save()` from inside `u3_pier_snap()` if the event is already both computed and committed.